Development

mikloslorinczi opened the pull request:
Alpha version working.

Infra CLI, Client and Server are three different app now.

Infra Server uses gorilla/mux router to serve REST JSON endpoits. It hosts the Task database which can be manipulated via the CLI.

The Infra Client regularly polls the server for new tasks to execute. When it finds one with a matching tags, it tries to execute the command and send the logfile to the server. Users can download these logs with the CLI.

Commit:
separated syncmap seeder and main using interface

Commit:
safeMap fixed, 100% test coverage

Commit:
function names updated

Commit:
added ListTasks and AddTask to cli, working with mock server

Commit:
infraserver created with JSON db, open/save working

Commit:
minor fix according to gometalinter

Commit:
complet refact! separated apps: cli, client and server

Commit:
Client name added

Commit:
Major simplify

Commit:
minor cosmetics according to metalinter

Commit:
Client polls the server, executes tasks

Commit:
Client print infromation on startup

Commit:
download log added, Alpha version working

Commit:
Readme.md updated, added docs to logs/ folders

Commit:
minor refact

Commit:
build.sh added, envFile added, db moved to own package

Commit:
disabled gosec in server.go:68

Commit:
added some comments to client.go

Commit:
changed Docker image from golang to alpine

tamaseger81 commented on build.sh:
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
```
Ideally, this would be implemented as bitrise workflows as part of the bitrise.yml.

tamaseger81 commented on build.sh:
```diff
@@ -0,0 +1,19 @@
+echo "Building for Docker..."
+echo
+echo "GOOS set to Linux..."
+export GOOS=linux
```
It's better to autodetect it. By the other hand, the base docker image will determinate it anyway

tamaseger81 commented on cli/cli.go:
```diff
@@ -0,0 +1,109 @@
+)
+
+// This function has complexity 13 according to gocyclo, despite it is "perfectly" flat
+func main() { // nolint: gocyclo
```
// nolint: gocyclo

Cyclomatic complexity is very important, should not be ignored. It always implies unnecessary complex blocks those must be broken down into smaller pieces

tamaseger81 commented on cli/cli.go:
```diff
@@ -0,0 +1,109 @@
+	if list {
+		tasks, err := listTasks()
+		if err != nil {
+			fmt.Println(err)
```
`log.Fatal` does exactly the same, yields the error and call system exit

tamaseger81 commented on cli/cli.go:
```diff
@@ -0,0 +1,109 @@
+		printUsage()
+		return false, 0, nil
+	}
+	common.AdminPass = os.Getenv("ADMIN_PASSWORD")
```
It's very error prone to mutate an exported package variable from another package.
`AdminPass` should initialised within an `init()` function located in the `common` package, especially when default value comes from an env variable.

tamaseger81 commented on cli/cli.go:
```diff
@@ -0,0 +1,109 @@
+	}
+	common.AdminPass = os.Getenv("ADMIN_PASSWORD")
+	flag.Parse()
+	if common.AdminPass == "" {
```
Same as above, Access details rather be wrapped as a struct and passed as parameter into context than being represented as exported package variables.

tamaseger81 commented on cli/cli.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+func printUsage() {
+	fmt.Println("\nInfra CLI")
```
Golang's Back-tick string preserves the whitespaces, it also supports variable substitution and fmt patterns. 

https://yourbasic.org/golang/multiline-string/

tamaseger81 commented on cli/handlers.go:
```diff
@@ -0,0 +1,126 @@
+	if err != nil {
+		return responseMsg, errors.Wrap(err, "Error reading commoand")
+	}
+	err = common.ToJSON(&commandObj, &commandJSON)
```
Variable `err` can be made local to the conditional block easily here, e.g.:

```
if err := common.ToJSON(&commandObj, &commandJSON); err != nil {
		return responseMsg, errors.Wrap(err, "Cannot add new task")
}
```

tamaseger81 commented on cli/handlers.go:
```diff
@@ -0,0 +1,126 @@
+	if err != nil {
+		return task, errors.Wrap(err, "Cannot query task")
+	}
+	err = common.FromJSON(&task, taskJSON)
```
Variable err can be made local to the conditional block easily here, e.g.:
```
if err :=...; err != nil {
		...
}
```

tamaseger81 commented on cli/handlers.go:
```diff
@@ -0,0 +1,126 @@
+	if !strings.Contains(task.Status, "log available") {
+		return fmt.Errorf("No log available for Task: %v Status: %v", task.ID, task.Status)
+	}
+	err = downloadLog(task.ID)
```
Variable err can be made local to the conditional block easily here, e.g.:
```
if err :=...; err != nil {
		...
}
```

tamaseger81 commented on cli/handlers.go:
```diff
@@ -0,0 +1,126 @@
+		}
+	}()
+
+	_, err = io.Copy(outFile, resp.Body)
```
Variable err can be made local to the conditional block easily here, e.g.:
```
if err :=...; err != nil {
		...
}
```

tamaseger81 commented on cli/handlers.go:
```diff
@@ -0,0 +1,126 @@
+		return errors.Wrap(err, "Error requesting logfile")
+	}
+	defer func() {
+		closeErr := resp.Body.Close()
```
Variable err can be made local to the conditional block easily here, e.g.:
```
if err :=...; err != nil {
		...
}
```

tamaseger81 commented on client/client.go:
```diff
@@ -0,0 +1,145 @@
+package main
```
I would give the same comments as for the `cli/cli.go`

- Cyclomatic complexity has to be kept minimal, must not being ignored
- Misused exposed package variables
- Multiline strings
- os.Env must be consumed within `init()` functions where it's possible
- error variables and checks would be kept short for conditional blocks where is possible, e.g.: 
```
if err :=...; err != nil {
		...
}
```

tamaseger81 commented on client/client.go:
```diff
@@ -0,0 +1,145 @@
+
+	for running := true; running; {
+		// Calculate next tick
+		nextTick := time.Now().Add(time.Second * time.Duration(period))
```
Tickers should be closed in defer blocks at any case.

https://golang.org/pkg/time/#example_NewTicker

tamaseger81 commented on client/client.go:
```diff
@@ -0,0 +1,145 @@
+		// Polling Infra Server for Tasks
+		fmt.Println("Polling Infra Server...")
+		task, ok, err1 := getTaskToExec()
+		if err1 != nil || !ok {
```
way too complex, why is there an `ok` and an `error` together?
How should an operation be `ok` when it returns an `error`? 
I quess the variable `ok` is spare here...

mikloslorinczi replied:
In case there is no task to execute OK will be false, but this doesn't indicate an error (or should it be an error?)

tamaseger81 commented on client/client.go:
```diff
@@ -0,0 +1,145 @@
+			waitNext(nextTick)
+			continue
+		}
+		// Try to claim matching Task
```
Couldn't these subtasks being implemented as separate functions?

tamaseger81 commented on client/handlers.go:
```diff
@@ -0,0 +1,154 @@
+				found = true
+			}
+		}
+		if !found {
```
`return found` ?
What is the reason of having an extra `if` block here?

tamaseger81 replied:
Should be easier to apply a short cut logic here? e.g.:
```
        taskSplit := strings.Split(taskTags, " ")
	nodeSplit := strings.Split(nodeTags, " ")
         # slices above would be sorted alphabetically as well 
         # and transformed to either lower or upper case for a case insensitive use-case
	for _, taskTag := range taskSplit {
		for _, nodeTag := range nodeSplit {
			if taskTag == nodeTag {
				return true
			}
		}
	}
	return false
```

mikloslorinczi replied:
If I understand correctly all of the Task's tags needs to be matched, not just the first.

tamaseger81 commented on client/handlers.go:
```diff
@@ -0,0 +1,154 @@
+	unassigned := []common.Task{}
+	for _, task := range tasks {
+		if task.Node == "None" {
+			found = true
```
Why don't you just return the values from here?
I neither see a reason to continue the loop nor the carry the values when a match has been found.

mikloslorinczi replied:
This function aims to select all Tasks which is unassigned (task.Node == "None"), so they can be checked if the Node's Tags are matching with any of them.

tamaseger81 commented on client/handlers.go:
```diff
@@ -0,0 +1,154 @@
+		responseJSON []byte
+		taskJSON     []byte
+	)
+	err := common.ToJSON(&task, &taskJSON)
```
err check short form could be used here,
```
if err := ... ; err != nil {}
```

tamaseger81 commented on client/handlers.go:
```diff
@@ -0,0 +1,154 @@
+		return responseMsg, errors.Wrap(err, "Cannot update task")
+	}
+	err = common.FromJSON(&responseMsg, responseJSON)
+	if err != nil {
```
err check short form could be used here

tamaseger81 commented on client/handlers.go:
```diff
@@ -0,0 +1,154 @@
+	}
+
+	defer func() {
+		err = outputFile.Close()
```
err check short form could be used here

tamaseger81 commented on client/handlers.go:
```diff
@@ -0,0 +1,154 @@
+	if err != nil {
+		return errors.Wrap(err, "Cannot open logfile")
+	}
+	// File already closed. Why?
```
Dead code?

tamaseger81 commented on client/handlers.go:
```diff
@@ -0,0 +1,154 @@
+		}
+	}()
+
+	_, err = ioutil.ReadAll(resp.Body)
```
err check short form could be used here

tamaseger81 commented on client/handlers_test.go:
```diff
@@ -0,0 +1,36 @@
+
+func (s *ClientHandlerTestSuite) TestCompareTags_Empty() {
+	match := compareTags("", "")
+	require.Equal(s.T(), true, match, "compareTags should return ture when fed with two empty strings")
```
The `ClientHandlerTestSuite` inherits from the `require.Assertions`, therefore the function must use it's own receivers exports e.g.:
`s.Equal(true, match, "compareTags should return ture when fed with two empty strings")`

tamaseger81 replied:
For `Equal(true, ...` matchers there's a built'n forwarded matcher `True(...)`

tamaseger81 commented on client/handlers_test.go:
```diff
@@ -0,0 +1,36 @@
+
+func (s *ClientHandlerTestSuite) TestCompareTags_Different() {
+	match := compareTags("a", "b")
+	require.Equal(s.T(), false, match, "compareTags should return false when fed with two completly different strings")
```
Same as above, `s.False(....)` would be the correct form here

tamaseger81 commented on client/handlers_test.go:
```diff
@@ -0,0 +1,36 @@
+
+func (s *ClientHandlerTestSuite) TestCompareTags_Matching() {
+	match := compareTags("cat dog penguin elephant shark", "bird dog worm cat penguin crocodile shark elephant")
+	require.Equal(s.T(), true, match, "compareTags should return ture when the second string contains all tag from the first string")
```
Same as above `s.True(....)`

tamaseger81 commented on client/logs/bdqmlv3oo3suave89mr0.log:
```diff
@@ -0,0 +1 @@
+sajt
```
I don't think these files supposed to be under version control. Missing a `.gitignore` directive?

tamaseger81 commented on client/logs/bdqmmm3oo3suave89mrg.log:
```diff
@@ -0,0 +1,4 @@
+client.go
```
I don't think these files supposed to be under version control. Missing a `.gitignore` directive?

tamaseger81 commented on client/logs/bdqmnrroo3suave89msg.log:
```diff
@@ -0,0 +1,22 @@
+
```
I don't think these files supposed to be under version control. Missing a `.gitignore` directive?

tamaseger81 commented on client/logs/bdqmpmboo3sunlgvprdg.log:
```diff
@@ -0,0 +1 @@
+"hi yo hejho"
```
I don't think these files supposed to be under version control. Missing a `.gitignore` directive?

tamaseger81 commented on client/logs/bdqmpvjoo3sunlgvpre0.log:
```diff
@@ -0,0 +1,18 @@
+Applications
```
I don't think these files supposed to be under version control. Missing a `.gitignore` directive?

tamaseger81 commented on common/loadenv.go:
```diff
@@ -0,0 +1,41 @@
+	"github.com/pkg/errors"
+)
+
+// LoadEnv accepts an envfile (with lines like: key=value) as parameter,
```
`JSON` and `YAML` are preferred over property files

tamaseger81 commented on common/structs.go:
```diff
@@ -0,0 +1,97 @@
+)
+
+// JSONable all struct which implements the marshal and unmarshal methods, to convert it to JSON and back.
+type JSONable interface {
```
There's a built'n interface for both serialisation and deserialisation which would be implemented by structs those require custom serialisation logic for any reason.

http://choly.ca/post/go-json-marshalling/

tamaseger81 commented on db/db.go:
```diff
@@ -0,0 +1,157 @@
+)
+
+// I is the interface of the JSON DB.
+type I interface {
```
Name `I` does not obey to the go naming conventions.

https://golang.org/doc/effective_go.html#interface-names

tamaseger81 commented on db/db.go:
```diff
@@ -0,0 +1,157 @@
+}
+
+// ConnectJSONDB returns a pointer to a jsonDB, working with the given path.
+func ConnectJSONDB(path string) I {
```
It looks like a constructor to me, therefore it should named as `New...`

tamaseger81 commented on db/db_test.go:
```diff
@@ -0,0 +1,87 @@
+
+func (s *DBTestSuite) TestQueryAll_EmptyDB() {
+	tasks, found := s.TestDB.QueryAll()
+	require.Equal(s.T(), []common.Task(nil), tasks, "QueryAll shoud return an empty slice of Tasks")
```
Why u use the global `require` functions when the receiver inherits the `Assertations`?

`s.Equal([]common.Task(nil), tasks, "QueryAll shoud return an empty slice of Tasks")`

tamaseger81 commented on db/db_test.go:
```diff
@@ -0,0 +1,87 @@
+
+func (s *DBTestSuite) TestAdd_Empty_Task() {
+	id, err := s.TestDB.Add(common.Task{})
+	require.NoError(s.T(), err, "Add shouldn't return an error when fed with an empty Task")
```
`s.NoError`

tamaseger81 commented on db/db_test.go:
```diff
@@ -0,0 +1,87 @@
+
+func (s *DBTestSuite) TestAdd_Task() {
+	id, err := s.TestDB.Add(common.Task{Command: "echo cheese", Tags: "tag1 tag2 tag3"})
+	require.NoError(s.T(), err, "Add shouldn't return an error when fed with a valid Task")
```
same as above

tamaseger81 commented on db/db_test.go:
```diff
@@ -0,0 +1,87 @@
+	id, err := s.TestDB.Add(common.Task{Command: "echo cheese", Tags: "tag1 tag2 tag3"})
+	require.NoError(s.T(), err, "Add shouldn't return an error when fed with a valid Task")
+	task, ok := s.TestDB.Query(id)
+	require.Equal(s.T(), true, ok, "Query should return OK true, if Task is found")
```
same as above

tamaseger81 commented on db/db_test.go:
```diff
@@ -0,0 +1,87 @@
+	require.Equal(s.T(), "echo cheese", task.Command, "Query should return the correct Task")
+}
+
+func (s *DBTestSuite) TestRemove_Task() {
```
`s.Equal`, `s.True`, etc...

tamaseger81 commented on db/db_test.go:
```diff
@@ -0,0 +1,87 @@
+}
+
+func (s *DBTestSuite) TestUpdate_Task() {
+	id, err := s.TestDB.Add(common.Task{Command: "echo cheese", Tags: "tag1 tag2 tag3"})
```
same as above

tamaseger81 commented on iclient.Dockerfile:
```diff
@@ -0,0 +1,11 @@
+FROM alpine
```
The binary would be built with the target image, otherwise it won't be compatible with the execution env.

Always use the multi stage builder pattern of docker: 
https://docs.docker.com/develop/develop-images/multistage-build/

tamaseger81 commented on iserver.Dockerfile:
```diff
@@ -0,0 +1,11 @@
+FROM alpine
```
The binary would be built with the target image, otherwise it won't be compatible with the execution env.

Always use the multi stage builder pattern of docker:
https://docs.docker.com/develop/develop-images/multistage-build/

Commit:
Implemented Cobra and Viper

Commit:
test fixed

Commit:
content of build.sh moved to bitrise.yml

mikloslorinczi commented:
- Cobra implemented to handle commandline arguments and help in program structure
- Viper implemented to handle config.yaml files and ENV variables. (unset < default < yaml < ENV < flag)
- Tests fixed to use "s." instead of "require."
- Short error handling implemented where it was possible.
- Logs removed from repo, added to .gitignore
- Fixed Dockerfiles, using the multi-stage build pattern now (golang as builder, alpine as final OS)
- Cyclomatic complexity is <= 10 in every function now (removed nolint)

Commit:
test

Commit:
testing tags

Commit:
additional taging test 1

Commit:
additional taging test 2

Commit:
additional taging test 3

Commit:
additional taging test 4

Commit:
additional taging test 5

Commit:
additional taging test 6

Commit:
updated README.md
